Rock/Security/Authentication/Google.cs

```diff
@@ -0,0 +1,363 @@
+                    try { email = googleUser.email.ToString(); }
+                    catch { }
+
+                    Person person = null;
+
+                    // If person had an email, get the first person with the same name and email address.
+                    if ( !string.IsNullOrWhiteSpace(email) )
+                    {
+                        var personService = new PersonService(rockContext);
+                        var people = personService.GetByMatch(firstName, lastName, email);
+                        if ( people.Count() == 1 )
+                        {
+                            person = people.First();
+                        }
+                    }
+
+                    var personRecordTypeId = DefinedValueCache.Read(SystemGuid.DefinedValue.PERSON_RECORD_TYPE_PERSON.AsGuid()).Id;
+                    var personStatusPending = DefinedValueCache.Read(SystemGuid.DefinedValue.PERSON_RECORD_STATUS_PENDING.AsGuid()).Id;
+
+                    rockContext.WrapTransaction(( ) =>
+                    {
+                        if ( person == null )
+                        {
+                            person = new Person();
+                            person.IsSystem = false;
+                            person.RecordTypeValueId = personRecordTypeId;
+                            person.RecordStatusValueId = personStatusPending;
+                            person.FirstName = firstName;
+                            person.LastName = lastName;
+                            person.Email = email;
+                            person.IsEmailActive = true;
+                            person.EmailPreference = EmailPreference.EmailAllowed;
+                            try
+                            {
+                                if ( googleUser.gender.ToString() == "male" )
+                                {
+                                    person.Gender = Gender.Male;
+                                }
+                                else if ( googleUser.gender.ToString() == "female" )
+                                {
+                                    person.Gender = Gender.Female;
+                                }
+                                else
+                                {
+                                    person.Gender = Gender.Unknown;
+                                }
+                            }
+                            catch { }
+
+                            if ( person != null )
+                            {
+                                PersonService.SaveNewPerson(person, rockContext, null, false);
+                            }
+                        }
+
+                        if ( person != null )
+                        {
+                            int typeId = EntityTypeCache.Read(typeof(Google)).Id;
+                            user = UserLoginService.Create(rockContext, person, AuthenticationServiceType.External, typeId, userName, "goog", true);
```

Comment:
"goog" is just arbitrary. It's the password parameter for the UserLogin create method. It is unclear.